Updates runfabtests.sh for OSX, homebrew, and other little things.

sungeunchoi opened the pull request:
This commit gets the script working on OSX, with a goal of providing
an argument-less invocation that requires no set up.  This is intended
to be used as a quick check, as is used in the brew formula test I
have set up for fabtests.

- Allow the script to be run with no arguments.  This will keep us
  from having to fiddle with the brew test if we decide to change the
  way arguments are parsed, etc.  The default behavior is to use the
  sockets provider with loopback.

- Add loopback-only testing mode that simply runs the commands without
  ssh.  This allows someone to run the script without doing anything
  (e.g., setting up keys).

- Add arguments to mktemp that mimics the default behavior for linux

- Do not use the -r argument to xargs for OSX (default behavior already)

- Remove nanoseconds from the time string as %N is not supported on OSX

- If a test binary is not found or not executable, consider it "Notrun"

- Send cleanup output to /dev/null

- Fix usage

@bturrubiates @hppritcha @shefty 

FYI: At Howard's recommendation, I have made conforming homebrew formulas for libfabric and fabtests.  Most of these changes were necessary to be able to test the installation in a meaningful way.  It would be nice to get this all working for the SC tutorial so Mac people can easily install the latest version.  I have no idea how long it takes to get the formula accepted.

pmmccorm commented on scripts/runfabtests.sh:
```diff
@@ -2,23 +2,46 @@
+    XARGS_CMD="xargs"
+# else
+    # Could issue warning if we want
+fi
```
How about instead of trying to determine OS etc we just do something like: timeout=`which timeout` ?

sungeunchoi replied:
Great idea.

pmmccorm commented on scripts/runfabtests.sh:
```diff
@@ -2,23 +2,46 @@
-declare -r c_outp=$(mktemp)
-declare -r s_outp=$(mktemp)
+declare -r c_outp=$(mktemp $TMPDIR/fabtests.c_outp.XXXXXX)
+declare -r s_outp=$(mktemp $TMPDIR/fabtests.s_outp.XXXXXX)
```
Why do we need this? On my RHEL machine TMPDIR is not defined, what does OS X do differently?

sungeunchoi replied:
On reading the man pages, both versions use TMPDIR if it is defined, otherwise /tmp.  FWIW, TMPDIR is defined on my Mac.

pmmccorm replied:
So is it needed? We should probably be respecting $TMPDIR regardless.

sungeunchoi replied:
I'm not sure I understand your comment.  Up on line 22, I set TMPDIR to /tmp if it's not set.

pmmccorm replied:
What happens if you don't set TMPDIR at all?

sungeunchoi replied:
Ah, right.  I'm being dense.  I'll remove use of TMPDIR.

pmmccorm commented on scripts/runfabtests.sh:
```diff
@@ -191,18 +239,19 @@ function unit_test {
 	test_time=$(compute_duration "$start_time" "$end_time")
 
-	if [ $ret1 -eq 61 ]; then
+        local nr=$(not_run $ret1)
```
This just covers up real errors. Why is this a good idea?

sungeunchoi replied:
The "real errors" you're talking about are build errors, right?  That's a good point.  The motivation was to get around the fact that fi_msg_epoll is not built for OSX (no epoll).  The decision to not built it is made at configure time, and it seemed a shame to duplicate that logic in some form in this script.

If that's the only reasonable solution, then I can change it to do that, but I'm open to other ideas.

pmmccorm replied:
I think the fi_msg_epoll test should always be built, but on OSX (or wherever HAVE_EPOLL is not defined) msg_epoll.c should return -61 (notrun).

Or, does it need to be the linux specific epoll, why not just poll?

Either way hacking around it in the test automation is not the way to go.

shefty replied:
spaces were used instead of tabs here

sungeunchoi replied:
@shefty I had the wrong editing mode loaded.  Will fix this and others.

sungeunchoi replied:
@pmmccorm I agree, and let's fix it for v1.2.  In the meantime, I will make the brew test (which is written for 1.X, so as to get it in soon) very simple, with an explanation that for 1.2 it will be updated.

pmmccorm commented on scripts/runfabtests.sh:
```diff
@@ -2,23 +2,46 @@
+if [ "$OSNAME" == "linux" ]; then
+    declare -r sssh="timeout 60s ${bssh}"
+    declare -r lssh="timeout 90s ${bssh}"
+    XARGS_CMD="xargs -r"
```
We should also be able to work around the lack of -r without platform detection. We can just drop the "-r" and/or detect in the cleanup script when there are no running processes to kill.

sungeunchoi replied:
In that case, the piped command line fails, but we can ignore the resulting error.  Might make it harder to figure out if other parts of it failed, which probably is only going to be a problem when porting to new OS (e.g, Windows).

pmmccorm replied:
The exit trap can figure that out, ie:

```
PROGS=$(ps -eo comm,pid | grep '^fi_' | awk '{print \$2}')
if [ -z "$PROGS" ]; then
   ...
fi
```

sungeunchoi replied:
This would require ssh'ing two times if the host is remote.  Is that better or worse than ignoring a real error, which probably would only happen if this were taken to another platform?  I'm inclined to take your original suggestion and just drop the -r.

pmmccorm replied:
Yes good point.

a-ilango commented on scripts/runfabtests.sh:
```diff
@@ -2,23 +2,46 @@
 trap cleanup_and_exit SIGINT
 
+#
+# Default behavior with no args will use sockets provider with loopback
```
It would be good to add this info in the usage function too.

sungeunchoi replied:
Good idea.

shefty commented:
As general feedback, it is helpful if separate changes go into separate patches.  I wouldn't break this patch up at this point, but please keep this in mind for future changes.

shefty commented on scripts/runfabtests.sh:
```diff
@@ -249,10 +298,12 @@ function cs_test {
 
-	if [ $ret1 -eq 61 -a $ret2 -eq 61 ]; then
+        local nr1=$(not_run $ret1)
+        local nr2=$(not_run $ret2)
```
spaces were used instead of tabs

shefty commented on scripts/runfabtests.sh:
```diff
@@ -366,12 +417,26 @@ done
+    SSH_CLIENT="${ssh} ${CLIENT}"
+else
+    SSH_CLIENT=eval
+fi
```
Indentation here looks different than other portions of the script.

pmmccorm commented:
Also on a higher level: are all these changes worth it for:

"Allow the script to be run with no arguments. This will keep us
from having to fiddle with the brew test if we decide to change the
way arguments are parsed, etc. The default behavior is to use the
sockets provider with loopback."

The arguments have changed once or twice (in compatible fashion) over the lifetime of the script. It might just be easier to make a scripts/brewtest.sh that does exactly what you want. But is sockets-loopback worth it?

shefty commented:
I run with the sockets provider in loopback *all* the time; it's useful when modifying the fabtests.  So, from my viewpoint, this is useful, and a better option than adding a new script so that I don't need to type 'sockets 127.0.0.1 127.0.0.1'.

sungeunchoi commented:
Using the same script also makes one less place to maintain OSX.

sungeunchoi commented:
@shefty Since there are a number of changes to be made to this commit, I will separate out the OSX changes from the changes made for the homebrew formula.

sungeunchoi commented:
@shefty @pmmccorm @a-ilango 

I've updated based on your comments.  I've also separated into 2 commits (one for OSX changes, one for homebrew) and fixed tabbing, etc.

pmmccorm commented on scripts/runfabtests.sh:
```diff
@@ -366,12 +379,26 @@ done
+	SERVER=$2
+	SSH_SERVER="${ssh} ${SERVER}"
+else
+	SSH_SERVER=eval
```
nit: just set ssh_{client,server} default values (eval) at the top and skip both of these else conditions.

pmmccorm commented on scripts/runfabtests.sh:
```diff
@@ -2,23 +2,35 @@
-declare -r sssh="timeout 60s ${bssh}"
-declare -r lssh="timeout 90s ${bssh}"
+declare TIMEOUT=`which timeout`
+if [ -z "$TIMEOUT" ]; then
```
nit: simplify and suppress error message like: `if [ -z $(which timeout 2> /dev/null) ]; then`

pmmccorm commented:
Small nits, but otherwise looks good to me.

a-ilango commented:
:+1:

sungeunchoi commented:
@pmmccorm Addressed your latest comments.

pmmccorm commented:
Thanks, looks good to me.

jithinjosepkl commented:
@sungeunchoi - thanks for the PR. 

Can we exclude msg_epoll test (at least for now) from the "simple_tests" list for osx?  This test is not built on osx.

sungeunchoi commented:
@pmmccorm suggesting building all tests and returning error if it doesn't work on a particular platform.  Then the script can check for that error.  I like the idea, but we will have to deal with different values for return codes (actually have to deal with that anyways, but ignored it in this PR).  I plan to open an issue about it for discussion.

jithinjosepkl commented:
ok, sounds good.
